browser/ai_chat/content_agent_tool_provider.cc

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ai_chat/content_agent_tool_provider.h"
+
+#include <memory>
+#include <vector>
+
+#include "brave/browser/ai_chat/tools/navigation_tool.h"
+#include "brave/components/ai_chat/core/browser/tools/tool.h"
+#include "brave/components/ai_chat/core/browser/tools/tool_provider.h"
+#include "brave/components/ai_chat/core/browser/tools/tool_utils.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom.h"
+#include "chrome/browser/actor/actor_keyed_service.h"
+#include "chrome/browser/actor/actor_task.h"
+#include "chrome/browser/actor/browser_action_util.h"
+#include "chrome/browser/actor/task_id.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser_navigator.h"
+#include "chrome/browser/ui/browser_tabstrip.h"
+#include "chrome/browser/ui/browser_window/public/browser_window_interface.h"
+#include "chrome/browser/ui/browser_window/public/browser_window_interface_iterator.h"
+#include "chrome/common/actor.mojom.h"
+#include "components/optimization_guide/content/browser/page_content_proto_provider.h"
+#include "components/tabs/public/tab_interface.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/base/page_transition_types.h"
+
+namespace ai_chat {
+
+ContentAgentToolProvider::ContentAgentToolProvider(
+    Profile* profile,
+    raw_ptr<actor::ActorKeyedService> actor_service)
+    : actor_service_(actor_service), profile_(profile) {
+  // Only create page actor tools if this profile is allowed to.
+  if (actor_service_) {
+    // Each conversation can have a different actor service task,
+    // and operate on a different set of tabs.
+    // If we want to delay creation of the task, we'll need to perhaps
+    // intercept all tool use calls and create or choose which task to use
+    // at that time. Tool::UseTool will have to change to
+    // ToolProvider::UseTool, or similar.
+    task_id_ = actor_service_->CreateTask();
+  }
+  CreateTools();
+}
+
+ContentAgentToolProvider::~ContentAgentToolProvider() = default;
+
+std::vector<base::WeakPtr<Tool>> ContentAgentToolProvider::GetTools() {
+  // Note: We don't have the ability to filter tools based on conversation
+  // capability here. But for now we don't need to as we only create the content
+  // agent tools if we don't have the actor_service_ (which is only provided for
+  // agent profiles).
+  std::vector<base::WeakPtr<Tool>> tool_ptrs;
+  tool_ptrs.reserve(tools_.size());
+  for (const auto& tool : tools_) {
+    tool_ptrs.push_back(tool->GetWeakPtr());
+  }
```

Comment:
optional: can we do a std::ranges::transform here with `&Tool::GetWeakPtr` as the transform function?